File: src/day1/go/binary_search_list.go
package main

func BinarySearchList(haystack []int, needle int) bool {
	lo := 0
	hi := len(haystack)

	for lo < hi {
		mid := lo + (hi-lo)/2

		switch {
		case haystack[mid] == needle:
			return true
		case haystack[mid] < needle:
			lo = mid + 1
		case haystack[mid] > needle:
			hi = mid
		}
	}
	return false
}

File: src/day1/python/binary_search_list.py
def binary_search_list(haystack: list[int], needle: int) -> bool:
    hi = len(haystack)
    lo = 0

    while lo < hi:
        MID = int(lo+(hi-lo)/2)

        if haystack[MID] == needle:
            return True

        if haystack[MID] < needle:
            lo = MID+1
        else:
            hi = MID

    return False
